### PR TITLE
feat(ci): smart upstream-monitor with release intelligence + path-aware CodeRabbit profile

### LINKED ISSUE #126: feat(ci): smart upstream-monitor with release intelligence
## Summary

Current `upstream-monitor.yml` is a dumb sync — it counts commits behind upstream and tries to merge. It provides zero intelligence about **what** changed upstream. We need feature-level visibility to make informed merge decisions.

## Current behavior

| Feature | Status |
|---------|--------|
| Detect new upstream commits | yes |
| Auto-merge if clean | yes |
| Create conflict issue | yes |
| Detect merged feature branches | yes |
| **Parse upstream releases/tags** | **no** |
| **Show what changed (features/fixes/breaking)** | **no** |
| **Link to upstream issues/PRs** | **no** |
| **Detect breaking changes** | **no** |
| **Categorize by conventional commits** | **no** |

### 1. Release detection

Check upstream tags/releases, not just commits. When a new release appears:

```
🏷️ New upstream release: lsm-tree v3.2.0 (was v3.1.2)
```

### 2. Commit categorization

Parse `git log` between our last sync point and upstream HEAD using conventional commit format:

```markdown
## Upstream changes since last sync

### ⚠️ Breaking (1)
- `feat!: remove deprecated flush_sync()` — fjall-rs/lsm-tree#290

### ✨ Features (3)
- `feat: add custom merge operator support` — fjall-rs/lsm-tree#280
- `feat(bloom): partitioned bloom filters` — fjall-rs/lsm-tree#275
- `feat: io_uring backend` — fjall-rs/lsm-tree#270

### 🐛 Fixes (2)
- `fix: race condition in concurrent compaction` — fjall-rs/lsm-tree#285
- `fix(vlog): corrupted blob header on crash` — fjall-rs/lsm-tree#282

### ⚡ Performance (1)
- `perf: vectorized block decoding` — fjall-rs/lsm-tree#278

### 📝 Other (4)
- `docs: update MSRV to 1.91` — fjall-rs/lsm-tree#288
- `test: add property tests for range scan` — fjall-rs/lsm-tree#286
- `chore(deps): bump lz4_flex to 0.14` — fjall-rs/lsm-tree#284
- `ci: add aarch64 cross-compilation` — fjall-rs/lsm-tree#283
```

### 3. Issue/PR linking

Extract `#NNN` references from commit messages and link them:

```bash
git log origin/main..upstream/main --format="%s" | grep -oE '#[0-9]+' | sort -u
```

For each reference, fetch title from upstream repo API.

### 4. Fork overlap detection

Check if any upstream changes touch files that our fork patches have modified:

```bash
# Files we've changed vs upstream
OUR_FILES=$(git diff origin/main...upstream/main --name-only)
FORK_FILES=$(git log --all --format="" --name-only -- src/ | sort -u)
OVERLAP=$(comm -12 <(echo "$OUR_FILES" | sort) <(echo "$FORK_FILES" | sort))
```

If overlap exists → flag as "needs manual review" even if merge is clean (semantic conflicts possible).

### 5. Smart PR/issue body

Instead of generic "N new commits", generate a structured body:

```markdown
## Upstream Sync: v3.1.2 → v3.2.0

**Release:** [v3.2.0](https://github.com/fjall-rs/lsm-tree/releases/tag/v3.2.0)
**Commits:** 47 new commits since last sync
**Breaking changes:** 1 ⚠️

### Changes by category
[categorized list from step 2]

### Fork overlap
These upstream changes touch files our fork has modified:
- `src/compaction/worker.rs` — our merge operator patches may need adaptation
- `src/tree/mod.rs` — our prefix bloom integration

### Review checklist
- [ ] Breaking changes evaluated for fork impact
- [ ] Overlapping files reviewed for semantic conflicts
- [ ] Fork-specific tests pass with upstream changes
- [ ] lsm-tree version in fjall Cargo.toml updated
```

### New workflow steps

```yaml
- name: Analyze upstream changes
  id: analyze
  run: |
    # 1. Detect release
    LATEST_TAG=$(git tag -l --sort=-v:refname 'v*' | head -1)
    UPSTREAM_TAG=$(git ls-remote --tags upstream | grep -oE 'v[0-9]+\.[0-9]+\.[0-9]+$' | sort -V | tail -1)

    # 2. Categorize commits
    git log origin/main..upstream/main --format="%s" | while read msg; do
      case "$msg" in
        feat\!:*|*BREAKING*) echo "breaking: $msg" ;;
        feat:*|feat\(*) echo "feature: $msg" ;;
        fix:*|fix\(*) echo "fix: $msg" ;;
        perf:*|perf\(*) echo "perf: $msg" ;;
        *) echo "other: $msg" ;;
      esac
    done > /tmp/categorized.txt

    # 3. Extract issue references
    git log origin/main..upstream/main --format="%s %b" | grep -oE '#[0-9]+' | sort -un > /tmp/refs.txt

    # 4. Check fork overlap
    git diff --name-only origin/main..upstream/main > /tmp/upstream_files.txt
    # Compare with our fork-specific changes (commits not in upstream)
    git diff --name-only upstream/main..origin/main > /tmp/fork_files.txt
    comm -12 <(sort /tmp/upstream_files.txt) <(sort /tmp/fork_files.txt) > /tmp/overlap.txt
```

### Backward compatible

- Keep existing auto-merge/conflict behavior
- Add intelligence as **additional context** in PR body / issue body
- No change to schedule (Mon/Thu 8am UTC)

## Acceptance criteria

- [ ] Detects new upstream release tags (not just commits)
- [ ] Categorizes commits by conventional commit type
- [ ] Extracts and resolves upstream issue/PR references (title + URL)
- [ ] Detects fork file overlap (potential semantic conflicts)
- [ ] Generates structured PR body with all of the above
- [ ] Generates structured issue body (conflict case) with all of the above
- [ ] Breaking changes highlighted prominently
- [ ] Backward compatible with current merge/conflict behavior

## Related

- fjall will get the same upgrade (separate issue)
- gitlab-mcp needs upstream-monitor from scratch (separate issue)
- strongswan already has sync-upstream.yml — evaluate if it needs the same intelligence

## Time estimate

1d — shared script + workflow updates for lsm-tree (fjall is copy+adapt)

---

## Additional: Add `.coderabbit.yaml` configuration

Currently no `.coderabbit.yaml` exists — CodeRabbit uses defaults. Add a proper config to:
- Disable pauses (reviews should never be paused/skipped)
- Set assertive profile (systems code needs thorough review)
- Add path-specific instructions (Rust safety, tests, benchmarks)
- Enable knowledge base with our coding guidelines
- Disable poem (noise in review walkthrough)
- Enable related issues/PRs detection

### Proposed `.coderabbit.yaml`

```yaml
# yaml-language-server: $schema=https://coderabbit.ai/integrations/schema.v2.json
language: "en-US"
early_access: true
reviews:
  profile: "assertive"
  request_changes_workflow: false
  high_level_summary: true
  high_level_summary_placeholder: "@coderabbitai summary"
  poem: false
  review_status: true
  collapse_walkthrough: true
  commit_status: true
  assess_linked_issues: true
  related_issues: true
  related_prs: true
  suggested_labels: true
  auto_review:
    enabled: true
    auto_incremental_review: true
    drafts: false
    base_branches: ["main"]
    ignore_title_keywords: ["WIP", "DO NOT MERGE"]
  path_instructions:
    - path: "src/**/*.rs"
      instructions: |
        Review for memory safety, ownership correctness, and panic-free error handling.
        No unwrap() on I/O or user-input paths. Prefer Result<T, E> everywhere.
        No Box<dyn Any> as type bypass. No global mutable state (lazy_static Mutex<HashMap>).
        Check multi-instance safety: no in-memory-only mutable state that breaks with N replicas.
    - path: "tests/**"
      instructions: |
        Verify test coverage of edge cases and error paths.
        No unwrap() on paths that test error handling.
        Tests must have descriptive names and comments explaining WHAT is being tested.
    - path: "benches/**"
      instructions: |
        Check benchmark methodology. Use criterion properly.
        Measure P99/P999 latency, not just throughput.
    - path: "src/compaction/**"
      instructions: |
        Compaction is crash-safety critical. Every state mutation must be atomic.
        Verify no partial writes can corrupt on-disk state.
    - path: "src/encryption.rs"
      instructions: |
        Security-critical code. Review for timing attacks, nonce reuse, key handling.
        RNG must be CSPRNG. No hardcoded keys or IVs.
  tools:
    shellcheck:
      enabled: true
    markdownlint:
      enabled: true
    github-checks:
      enabled: true
      timeout_ms: 120000
chat:
  auto_reply: true
knowledge_base:
  learnings:
    scope: auto
  issues:
    scope: auto
  pull_requests:
    scope: auto
  code_guidelines:
    enabled: true
    filePatterns:
      - ".github/copilot-instructions.md"
      - ".github/instructions/*.instructions.md"
issue_enrichment:
  labeling:
    auto_apply_labels: true
    labeling_instructions:
      - label: "bug"
        instructions: "Code defect, incorrect behavior, crash, data corruption, wrong results"
      - label: "enhancement"
        instructions: "New feature, new API, new capability not previously available"
      - label: "performance"
        instructions: "Optimization, reduced allocations, faster path, benchmark improvement"
      - label: "refactor"
        instructions: "Code restructuring without behavior change — renames, extractions, trait threading"
      - label: "test"
        instructions: "New tests, test infrastructure, test helpers, flaky test fixes"
      - label: "ci"
        instructions: "CI/CD workflows, GitHub Actions, benchmarks pipeline, release automation"
      - label: "comparator"
        instructions: "UserComparator threading, custom key ordering, lexicographic vs comparator-aware"
      - label: "compaction"
        instructions: "LSM compaction logic, leveled/tiered strategy, L0/L1/L2 overlap, compaction picker"
      - label: "crash-safety"
        instructions: "Crash recovery, fsync ordering, atomic writes, WAL correctness, data durability"
      - label: "encryption"
        instructions: "Block encryption, AES-GCM, key management, nonce handling"
      - label: "fs-trait"
        instructions: "Filesystem abstraction, Fs trait, io_uring, per-level routing, StdFs"
      - label: "upstream-candidate"
        instructions: "Fix or feature that could be contributed back to fjall-rs upstream"
      - label: "fork-only"
        instructions: "Feature specific to CoordiNode fork — range tombstones, merge operators, prefix bloom, V4 format"
      - label: "upstream-sync"
        instructions: "Automated upstream synchronization — merge conflicts, release tracking"
```

### Key decisions

| Setting | Value | Why |
|---------|-------|-----|
| `profile` | assertive | Systems/storage code needs thorough review — "chill" misses too much |
| `poem` | false | Noise in walkthrough, wastes review tokens |
| `auto_incremental_review` | true | Review each push, not just first |
| `drafts` | false | Don't review drafts — they're WIP |
| `path_filters` | none needed | lsm-tree repo has no dirs to exclude (donor codebases live in coordinode workspace, not here) |
| `path_instructions` | Rust safety rules | Encode our engineering principles directly into review guidelines |
| `early_access` | true | Get new CodeRabbit features as they ship |
| `knowledge_base.code_guidelines` | copilot-instructions.md | CodeRabbit auto-reads this for review context |
| `commit_status` | true | Block merge until review completes |
| `assess_linked_issues` | true | Check if PR actually addresses linked issue |

### Checklist addition

- [ ] Create `.coderabbit.yaml` with config above
- [ ] Verify CodeRabbit picks it up on next PR (comment `@coderabbitai configuration` to confirm)

### PR DESCRIPTION
**Pull request by polaz**
## Summary

Two CI tooling files. Together they upgrade the fork's upstream-sync flow from "count commits and try a merge" to a Dependabot-style dependency-aware sync surface that respects maintainer decisions.

- **`.github/workflows/upstream-monitor.yml`** — full rewrite. Surfaces upstream changes as a structured PR (clean merge) or issue (conflict), categorises commits by Conventional Commits, anchors the report to a release tag when one is available, resolves referenced upstream issues/PRs to titles, flags fork-overlap files for manual review, and tracks "already considered" upstream HEADs so a closed sync PR isn't auto-recreated until upstream advances.
- **`.coderabbit.yaml`** — expanded from a 7-line stub to a full path-aware review profile: knowledge-base linked to the project's `copilot-instructions.md` + `instructions/*.instructions.md`, path-specific review focus for security/compaction/encryption areas, full label-enrichment catalog matching the repo's actual label set.

## What the maintainer sees on a sync run

Each run produces ONE artifact for the current upstream HEAD:

- **Clean merge** → sync PR with title `chore: sync upstream-<short> (N new commits)`, body containing release detection, conventional-commit categorisation (Breaking ⚠️ / Features ✨ / Fixes 🐛 / Performance ⚡ / Other 📝), referenced upstream issues/PRs with resolved titles, fork-overlap file list, and a review checklist.
- **Conflict** → issue with title `Upstream sync conflict for upstream-<short> (N new commits)` carrying the same context plus a manual-resolution snippet. This is the steady state for forks with permanent rename / cherry-pick divergences.
- **Pre-merge failure** (dirty tree, internal git error) → distinct issue title `Upstream sync merge failed without conflicts for upstream-<short>` so on-call investigates the workflow rather than chasing phantom conflicts.

## "Already considered" semantic

Maintainer interactions on the auto-PR/issue ARE the considered signal — no separate state file edit or tag bump needed:

- **Open** → review in progress, future runs skip via `gh search issues "upstream-<short> in:title label:upstream-sync"`.
- **Merged** → user accepted; origin/main advances; next run reads `BEHIND=0` naturally.
- **Closed without merge** → user said "no" for this upstream HEAD; closed PR/issue still satisfies the title search, future runs skip until upstream advances (new short SHA → no search match → new artifact surfaced).

`workflow_dispatch` input `force=true` bypasses the skip for explicit "re-surface anyway" runs.

## Other workflow behaviour locked in

- **State file** `.github/state/last-upstream-sync.txt` — written on each successful merge, lazily bootstrapped via `--merged origin/main ∩ --merged upstream/main` (intersection guarantees only upstream-origin tags qualify, fork's own releases excluded).
- **Idempotent label create** — `upstream-sync` label is `--force`-created at the top of every gh-creating step so a deleted/missing label doesn't abort the workflow.
- **Early skip gate** — skip-check runs before the expensive Build step (up to 20 upstream API calls per ref), short-circuiting cron noise.
- **Sync-branch naming** — `chore/upstream-sync-<date>-<run-number>` avoids same-UTC-day collisions on re-runs.
- **Body composition** — checklist appended last on both routes (success / conflict); conflict block inserted before checklist for top-to-bottom reading order.
- **GH_REPO** pinned at workflow env level so `gh` calls don't accidentally route to the upstream remote.
- **md_escape** covers `` ` `` `|` `[` `]` `(` `)` `<` `>` `*` `_` `#` so arbitrary upstream issue titles can't break the rendered markdown.
- **Pre-split refs cap** — at most 20 referenced issues/PRs rendered per sync report; the report's primary value is "upstream has N new, mergeable: yes/no" rather than exhaustive ref enumeration.

## `.coderabbit.yaml` deliverables

- `profile: assertive`, `poem: false`, `auto_incremental_review: true`, `drafts: false`, `early_access: false` (stable channel for reproducible reviews).
- `path_instructions` for `src/**/*.rs`, `tests/**`, `benches/**`, `src/compaction/**`, `src/encryption*.rs` + `src/encryption/**` (mirrored blocks with `⚠️ Keep in sync` trailer; brace expansion + YAML anchors rejected as schema-uncertain).
- `knowledge_base.code_guidelines.filePatterns` → `.github/copilot-instructions.md` + `.github/instructions/*.instructions.md`.
- `issue_enrichment.labeling.auto_apply_labels: true` with 14 labels matching this repo's actual catalog (`bug`, `enhancement`, `performance`, `refactor`, `test`, `ci`, `documentation`, `comparator`, `compaction`, `crash-safety`, `encryption`, `fs-trait`, `upstream-candidate`, `fork-only`).
- `base_branches: ["main"]` documented inline (sync PRs also target `main`).

## Test plan

- [x] YAML parses (both files)
- [x] yamllint clean (both files)
- [x] shellcheck clean (every `run: |` block)
- [x] Locally tested: NUL-record parser handles multi-line commit body with `BREAKING CHANGE:` footer (2 records in / 2 categorised); ref-tokeniser strips path/to/file#42 false-positives and captures `fjall-rs/lsm-tree#456.` cross-repo refs after trailing-`.` strip; conventional-commit regex anchors prevent false-positive breaking-change classification.
- [x] Live-tested end-to-end via 5 `workflow_dispatch` runs on this branch — caught and fixed: (a) `gh` routing through upstream remote → pinned `GH_REPO`; (b) `git push` of tag pointing at upstream rejected for workflow files → simplified mechanism to title-search (no tags); (c) skip-check ordering moved before expensive Build for API-quota savings.
- [x] Close-without-merge respected: closing the test sync issue → next `workflow_dispatch` correctly emits `Skip: N sync artifact(s) for upstream-<short> already exist`.

## Out of scope (tracked separately if revisited)

- `force=true` in cron mode (currently dispatch-only).
- Multi-branch base support (sync always targets `main`; widen `base_branches` if that changes).
- Pre-release upstream tags (semver regex intentionally stable-only).

Closes #126

## Summary by CodeRabbit

* **Chores**
  * Expanded CodeRabbit automation: richer review rules, path-specific instructions, auto-labeling, chat auto-reply, enabled review tools, and adjusted timeouts.
  * Enhanced upstream sync workflow: manual dispatch, persisted state, safer remote handling, structured upstream change reports, idempotent merge/PR flow, improved conflict/issue creation, and updated merged-branch cleanup checks.

[Review Change Stack](https://app.coderabbit.ai/change-stack/structured-world/coordinode-lsm-tree/pull/259?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: .coderabbit.yaml [hunk 1]
```diff
@@ -1,6 +1,108 @@
...
+  poem: false
+  review_status: true
+  collapse_walkthrough: true
+  commit_status: true
+  assess_linked_issues: true
+  related_issues: true
+  related_prs: true
+  suggested_labels: true
   auto_label:
     enabled: true
```
[the remark this answers is not in the document]
Comment (polaz) [reply .coderabbit.yaml]: @coderabbitai Fixed in e2210c59 — `reviews.auto_label` block dropped. `issue_enrichment.labeling.auto_apply_labels` (true) is the only auto-label config now.

File: .github/workflows/upstream-monitor.yml [hunk 2]
```diff
@@ -20,111 +26,278 @@ jobs:
...
+
           echo "Commits behind upstream: $BEHIND"
+          echo "Upstream latest tag: ${UPSTREAM_LATEST_TAG:-<none>}"
+          echo "Upstream tag in our main: ${UPSTREAM_PREV_TAG:-<none>}"
+          echo "New release available: $NEW_RELEASE"
 
-      - name: Try merge and create PR or issue
-        if: steps.check.outputs.behind > 0
+      - name: Build change report
+        if: steps.detect.outputs.behind > 0
```
[the remark this answers is not in the document]
Comment (polaz) [reply .github/workflows/upstream-monitor.yml]: @coderabbitai Fixed in e2210c59 — all three step conditions now use `steps.detect.outputs.behind != '0'`.

File: .github/workflows/upstream-monitor.yml [hunk 3]
```diff
@@ -20,111 +27,343 @@ jobs:
...
+            elif [[ "$SUBJECT" =~ $FIX_RE ]]; then
+              echo "$LINE" >> /tmp/sync/fix.txt
+            elif [[ "$SUBJECT" =~ $PERF_RE ]]; then
+              echo "$LINE" >> /tmp/sync/perf.txt
+            else
+              echo "$LINE" >> /tmp/sync/other.txt
+            fi
+          done < <(git log "origin/main..${UPSTREAM_REMOTE}/main" \
+                     --format='%H%x09%s%x09%b' -z \
+                     | tr '\0' '\n')
```
[the remark this answers is not in the document]
Comment (polaz) [reply .github/workflows/upstream-monitor.yml]: @coderabbitai Fixed in 711b4f85. Parser now reads NUL-delimited records via `read -r -d ''` and splits the three tab-separated fields with bash parameter expansion (`${RECORD%%$'\t'*}` etc.), so multi-line bodies stay intact in `$BODY` and don't bleed into the next iteration. Locally verified with a synthesised `SHA\tSUBJECT\tBODY\0` payload containing an internal newline + `BREAKING CHANGE:` footer — 1 record in → 1 categorisation, breaking detected via the body grep. No more `tr '\0' '\n'` step.

File: .github/workflows/upstream-monitor.yml [hunk 4]
```diff
@@ -20,111 +27,404 @@ jobs:
...
-          - [ ] Run full test suite
-          EOF
-          )" \
+              --title "chore: sync upstream (${BEHIND} new commits)" \
+              --body-file "$BODY_PATH" \
               --base main \
               --head "$SYNC_BRANCH"
           else
             CONFLICTS=$(git diff --name-only --diff-filter=U 2>/dev/null || true)
             git merge --abort
```
[the remark this answers is not in the document]
Comment (polaz) [reply .github/workflows/upstream-monitor.yml:576]: @coderabbitai Fixed in cba79752 — `git merge --abort` is now guarded with `2>/dev/null || true` plus a comment explaining the defensive intent.

File: .github/workflows/upstream-monitor.yml [hunk 5]
```diff
@@ -20,111 +27,445 @@ jobs:
...
+          # Fork overlap: files touched by upstream (since the last shared
+          # commit) AND by fork-specific commits (commits on origin/main that
+          # are NOT in upstream/main — excludes upstream commits we already
+          # merged in, which would otherwise drown out the signal).
+          MERGE_BASE=$(git merge-base origin/main "${UPSTREAM_REMOTE}/main")
+          git diff --name-only "$MERGE_BASE..${UPSTREAM_REMOTE}/main" \
+            | sort -u > /tmp/sync/upstream_files.txt
+          git log origin/main --not "${UPSTREAM_REMOTE}/main" \
+            --name-only --format= \
+            | grep -v '^$' | sort -u > /tmp/sync/fork_files.txt || true
```
[the remark this answers is not in the document]
Comment (polaz) [reply .github/workflows/upstream-monitor.yml]: @coderabbitai Fixed — fork-side walk now uses `${MERGE_BASE}..origin/main --not ${UPSTREAM_REMOTE}/main` so the overlap list only contains fork commits since the current shared base.

File: .github/workflows/upstream-monitor.yml [hunk 6]
```diff
@@ -20,111 +27,445 @@ jobs:
...
+            # Kept out of the merge commit so fork-only bookkeeping doesn't
+            # mix with upstream content. Lazy-bootstrap: first sync with
+            # behind > 0 writes the file; until then the `git tag --merged
+            # origin/main` inference in the detect step covers the gap.
+            # Runs with behind == 0 don't reach this step.
+            if [ -n "${UPSTREAM_LATEST_TAG}" ]; then
+              mkdir -p "$(dirname "$STATE_FILE")"
+              echo "$UPSTREAM_LATEST_TAG" > "$STATE_FILE"
+              git add "$STATE_FILE"
+              git commit -m "chore: bump upstream sync state to ${UPSTREAM_LATEST_TAG}"
```
[the remark this answers is not in the document]
Comment (polaz) [reply .github/workflows/upstream-monitor.yml]: @coderabbitai Fixed — state-bump `git commit` now guarded with `if ! git diff --cached --quiet`. Between-release runs (same tag re-written) skip the commit instead of aborting the step under `set -e`.

File: .github/workflows/upstream-monitor.yml [hunk 7]
```diff
@@ -20,111 +40,574 @@ jobs:
...
+          # Mark this exact upstream HEAD as "considered" — applies on
+          # BOTH branches (successful merge PR + conflict issue). For a
+          # fork with permanent divergences (renames, cherry-pick-only
+          # relationships), the conflict path is the steady state and
+          # without this tag every cron run would refile the same issue.
+          # The tag stays regardless of how the maintainer dispositions
+          # the surfaced artifact. `|| true` because a tag collision on
+          # an explicit force=true re-run is harmless.
+          git tag "$CONSIDERED_TAG" "${UPSTREAM_REMOTE}/main"
+          git push origin "$CONSIDERED_TAG" || true
```
[the remark this answers is not in the document]
Comment (polaz) [reply .github/workflows/upstream-monitor.yml]: @coderabbitai Fixed — `git tag -f` plus a refspec-scoped tag-override push. force=true re-runs no longer abort on a local tag collision after checkout brings the prior tag down.

File: .github/workflows/upstream-monitor.yml [hunk 8]
```diff
@@ -20,111 +40,623 @@ jobs:
...
-            if git merge-base --is-ancestor "$BRANCH_TIP" upstream/main 2>/dev/null; then
-              echo "Branch '$BRANCH' is fully merged into upstream/main"
+            if git merge-base --is-ancestor "$BRANCH_TIP" "${UPSTREAM_REMOTE}/main" 2>/dev/null; then
+              echo "Branch '$BRANCH' is fully merged into ${UPSTREAM_REMOTE}/main"
 
-              EXISTING=$(gh issue list --search "Upstream merged: $BRANCH" --state open --json number --jq 'length')
+              EXISTING=$(gh issue list \
+                --search "Upstream merged: $BRANCH" \
+                --state open --json number --jq 'length')
               if [ "$EXISTING" = "0" ]; then
```
[the remark this answers is not in the document]
Comment (polaz) [reply .github/workflows/upstream-monitor.yml:685]: @coderabbitai Fixed — `SEARCH_BRANCH="${BRANCH//\#/}"` strips the hash before `gh issue list --search`. Indexed issue title gets stripped too, so the match works.

**Comment by polaz** [comment]
@coderabbitai Re: refs cap not applied pre-split (yml:243-254 duplicate). Fixed — refs are now capped to REFS_CAP=20 immediately after `sort -u`, then split into bare/qualified files. The reader loops are simplified accordingly (no per-list cap, no "N more" footers). Total-of-20 semantic is fine for this workflow since refs are second-tier info; the primary report value is "upstream has N new commits, mergeable: yes/no".

File: .github/workflows/upstream-monitor.yml [hunk 9]
```diff
@@ -20,111 +40,609 @@ jobs:
...
+            echo "upstream_latest_tag=$UPSTREAM_LATEST_TAG"
+            echo "upstream_prev_tag=$UPSTREAM_PREV_TAG"
+            echo "upstream_head_short=$UPSTREAM_HEAD_SHORT"
+          } >> "$GITHUB_OUTPUT"
+
+          NEW_RELEASE=false
+          if [ -n "$UPSTREAM_LATEST_TAG" ] && [ "$UPSTREAM_LATEST_TAG" != "$UPSTREAM_PREV_TAG" ]; then
+            NEW_RELEASE=true
+          fi
+          echo "new_release=$NEW_RELEASE" >> "$GITHUB_OUTPUT"
```
[the remark this answers is not in the document]
Comment (polaz) [reply .github/workflows/upstream-monitor.yml:127]: @coderabbitai Declined as a design choice — for a fork-with-divergences flow, tag-only releases on an upstream HEAD we've already considered carry no actionable signal (the code under that HEAD is what we decided about; the tag is a downstream-ceremony artifact). Added code comment near NEW_RELEASE detection documenting the choice. Whatever release transition matters surfaces in the heading of the next BEHIND>0 sync.

File: .github/workflows/upstream-monitor.yml [hunk 10]
```diff
@@ -20,111 +40,654 @@ jobs:
...
+          else
+            # Intersect "tags reachable from origin/main" with "tags
+            # reachable from upstream/main" — guarantees the inferred
+            # tag is upstream-origin and excludes fork-local release
+            # tags (e.g., our own `v4.4.0`) that would otherwise be
+            # mistaken for a previous upstream sync point.
+            UPSTREAM_PREV_TAG=$(comm -12 \
+              <(git tag -l --merged "origin/main" | sort) \
+              <(git tag -l --merged "${UPSTREAM_REMOTE}/main" | sort) \
+              | grep -E "$SEMVER_RE" | sort -V | tail -1 || true)
```
[the remark this answers is not in the document]
Comment (polaz) [reply .github/workflows/upstream-monitor.yml:106]: @coderabbitai Fixed — the bootstrap fallback now triple-intersects with the upstream-remote tag list, mirroring the safety pattern UPSTREAM_LATEST_TAG already uses. A fork-local `v4.4.0` reachable from both main heads via back-merge can no longer survive the filter unless upstream itself publishes that tag name.

File: .github/workflows/upstream-monitor.yml [hunk 11]
```diff
@@ -20,111 +40,662 @@ jobs:
...
+          # All sync PR / issue titles include the `upstream-<short>`
+          # fragment that the skip-step search looks for. The fragment
+          # uses a dash (not `@` or `#`) so GitHub's search-text
+          # normaliser doesn't strip it.
+
+          # Branch name carries the date + run number so a second run on the
+          # same UTC day (manual `workflow_dispatch` after the cron run) gets
+          # a distinct name and doesn't collide with an existing remote branch.
+          SYNC_BRANCH="chore/upstream-sync-$(date +%Y%m%d)-${GITHUB_RUN_NUMBER}"
           git checkout -b "$SYNC_BRANCH" origin/main
```
[the remark this answers is not in the document]
Comment (polaz) [reply .github/workflows/upstream-monitor.yml:523]: @coderabbitai Fixed — `SYNC_BRANCH` now includes `${GITHUB_RUN_ATTEMPT}` as the last segment, so re-running a failed workflow run produces a distinct branch name instead of colliding with the prior attempt.

**Comment by polaz** [comment]
Closing in favour of a fresh PR for the same branch. After 9 review rounds and 70+ resolved threads, the UI is unusable as a review surface. Code state on `feat/#126` is unchanged; a new PR will be opened against the same head for a clean review history. No new changes — same commits, same body, same scope.
